controllers/nonadmin_controller.go

```diff
@@ -0,0 +1,193 @@
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	oadpv1alpha1 "github.com/openshift/oadp-operator/api/v1alpha1"
+	"github.com/openshift/oadp-operator/pkg/common"
+)
+
+const (
+	nonAdminPrefix            = "non-admin-"
+	controllerManager         = "controller-manager"
+	containerName             = nonAdminPrefix + "manager"
+	nonAdminControllerManager = common.OADPOperatorPrefix + nonAdminPrefix + controllerManager
+	controlPlaneKey           = "control-plane"
+)
+
+var (
+	controlPlaneLabel = map[string]string{
+		controlPlaneKey: nonAdminPrefix + controllerManager,
+	}
+	deploymentLabels = map[string]string{
+		"app.kubernetes.io/component":  "manager",
+		"app.kubernetes.io/created-by": common.OADPOperator,
+		"app.kubernetes.io/instance":   nonAdminPrefix + controllerManager,
+		"app.kubernetes.io/managed-by": "kustomize",
+		"app.kubernetes.io/name":       "deployment",
+		"app.kubernetes.io/part-of":    common.OADPOperator,
+	}
+)
+
+func (r *DPAReconciler) ReconcileNonAdminController(log logr.Logger) (bool, error) {
+	// TODO https://github.com/openshift/oadp-operator/pull/1316
+	dpa := oadpv1alpha1.DataProtectionApplication{}
+	if err := r.Get(r.Context, r.NamespacedName, &dpa); err != nil {
+		return false, err
+	}
+
+	nonAdminDeployment := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      nonAdminControllerManager,
+			Namespace: r.NamespacedName.Namespace,
+		},
+	}
+
+	// Delete (possible) previously deployment
+	if !r.checkNonAdminEnabled(&dpa) {
+		if err := r.Get(
+			// TODO use r.Context?
+			context.Background(),
+			types.NamespacedName{
+				Name:      nonAdminDeployment.Name,
+				Namespace: nonAdminDeployment.Namespace,
+			},
+			nonAdminDeployment,
+		); err != nil {
+			if k8serror.IsNotFound(err) {
+				return true, nil
+			}
+			return false, err
+		}
+
+		deleteOptionPropagationForeground := metav1.DeletePropagationForeground
+		if err := r.Delete(
+			// TODO use r.Context?
+			context.Background(),
+			nonAdminDeployment,
+			&client.DeleteOptions{PropagationPolicy: &deleteOptionPropagationForeground},
+		); err != nil {
+			r.EventRecorder.Event(
+				nonAdminDeployment,
+				corev1.EventTypeWarning,
+				"NonAdminDeploymentDeleteFailed",
+				fmt.Sprintf("Could not delete non admin controller deployment %s/%s: %s", nonAdminDeployment.Namespace, nonAdminDeployment.Name, err),
+			)
+			return false, err
+		}
+		r.EventRecorder.Event(
+			nonAdminDeployment,
+			corev1.EventTypeNormal,
+			"NonAdminDeploymentDeleteSucceed",
+			fmt.Sprintf("Non admin controller deployment %s/%s deleted", nonAdminDeployment.Namespace, nonAdminDeployment.Name),
+		)
+		return true, nil
+	}
+
+	operation, err := controllerutil.CreateOrUpdate(
+		r.Context,
+		r.Client,
+		nonAdminDeployment,
+		func() error {
+			nonAdminImage := r.getNonAdminImage(&dpa)
+			if len(nonAdminImage) == 0 {
```

Comment:
Here in the mutation function lets follow the similar pattern of `buildNonAdminDeploymen` and `customizeNonAdminDeploymnet` as we do for `velero` and node-agent`. This helps code readability, maintainability and uniforminity.

Reply:
I though about that, and I do agree it was more readable, but I changed because the way it was, it was triggering update event every minute

here how it was https://github.com/openshift/oadp-operator/pull/1370/commits/25319386d99cf4c9c1db285d679000d63eb8d489#diff-ece703e5852171fe350087ab8bd7d5e14c7b577a83664fadc7506cafd54117e2R136

But object in cluster was like that
```yaml
kind: Deployment
apiVersion: apps/v1
metadata:
  annotations:
    deployment.kubernetes.io/revision: '1'
  resourceVersion: '322794519'
  name: openshift-adp-non-admin-controller-manager
  uid: 45d11346-500e-464b-8343-497f1d7b3eca
  creationTimestamp: '2024-03-27T18:50:52Z'
  generation: 1
  managedFields: 'A LOT OF THINGS...'
  namespace: test-oadp-operator
  ownerReferences:
    - apiVersion: oadp.openshift.io/v1alpha1
      kind: DataProtectionApplication
      name: velero-sample
      uid: 30827073-a33c-4694-9e51-bb2b6754ced4
      controller: true
      blockOwnerDeletion: true
  labels:
    app.kubernetes.io/component: manager
    app.kubernetes.io/created-by: oadp-operator
    app.kubernetes.io/instance: non-admin-controller-manager
    app.kubernetes.io/managed-by: kustomize
    app.kubernetes.io/name: deployment
    app.kubernetes.io/part-of: oadp-operator
    control-plane: non-admin-controller-manager
spec:
  replicas: 1
  selector:
    matchLabels:
      control-plane: non-admin-controller-manager
  template:
    metadata:
      creationTimestamp: null
      labels:
        control-plane: non-admin-controller-manager
    spec:
      containers:
        - name: non-admin-manager
          image: 'quay.io/msouzaol/non-admin-controller:latest'
          resources: {}
          terminationMessagePath: /dev/termination-log
          terminationMessagePolicy: File
          imagePullPolicy: Always
      restartPolicy: Always
      terminationGracePeriodSeconds: 30
      dnsPolicy: ClusterFirst
      serviceAccountName: openshift-adp-non-admin-controller-manager
      serviceAccount: openshift-adp-non-admin-controller-manager
      securityContext: {}
      schedulerName: default-scheduler
  strategy:
    type: RollingUpdate
    rollingUpdate:
      maxUnavailable: 25%
      maxSurge: 25%
  revisionHistoryLimit: 10
  progressDeadlineSeconds: 600
status:
  observedGeneration: 1
  replicas: 1
  updatedReplicas: 1
  readyReplicas: 1
  availableReplicas: 1
  conditions:
    - type: Available
      status: 'True'
      lastUpdateTime: '2024-03-27T18:50:59Z'
      lastTransitionTime: '2024-03-27T18:50:59Z'
      reason: MinimumReplicasAvailable
      message: Deployment has minimum availability.
    - type: Progressing
      status: 'True'
      lastUpdateTime: '2024-03-27T18:50:59Z'
      lastTransitionTime: '2024-03-27T18:50:52Z'
      reason: NewReplicaSetAvailable
      message: >-
        ReplicaSet "openshift-adp-non-admin-controller-manager-7bd4c9f4f6" has
        successfully progressed.
```

way different then yaml struct we were sending

if we can both make code do not trigger update event every minute and use yaml struct, I am all in!

Reply:
you are getting update events every minute because you might be updating the deployment in the mutation function of `createorupdate` call :)

Reply:
yes, that was the case because of previous implementation

this way, it only updates deployment if needed